BigQuery Denormalized : Cover arrays only if they are nested

DoNotPanicUA opened the pull request:
## What

- [x]  [11109](https://github.com/airbytehq/airbyte/issues/11109) Stop surrounding any array. Do it only if we have an array of arrays.
- [x]  [14058](https://github.com/airbytehq/airbyte/issues/14058) Fix processing an array of DateTime values.
- [x]  [13634](https://github.com/airbytehq/airbyte/issues/13634) Enable new DAT tests.
- [x]  [5912](https://github.com/airbytehq/airbyte/issues/5912) Reopened.
- [x]  [14668](https://github.com/airbytehq/airbyte/issues/14668) Cover by tests the BigQuery inheritance limitation
- [x] [15363](https://github.com/airbytehq/airbyte/issues/15363) Workaround

## Recommended reading order
1. `DefaultBigQueryDenormalizedRecordFormatter.java`
2. `BigQueryUtils.java`
3. `BigQueryDenormalizedDestinationTest.java`
4. `BigQueryDenormalizedDestinationAcceptanceTest.java`

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery-denormalized

grishick commented on airbyte-integrations/connectors/destination-bigquery-denormalized/Dockerfile:
```diff
@@ -17,5 +17,5 @@ ENV ENABLE_SENTRY true
 COPY --from=build /airbyte /airbyte
 
-LABEL io.airbyte.version=1.1.11
+LABEL io.airbyte.version=2.0.0
```
why major version change?

DoNotPanicUA replied:
There is no backward compatibility with existing connectors. The structure of the destination will be different now.
I decided that by the version, I could highlight it. 
What would you recommend doing in this case?

grishick replied:
What will happen if a user upgrades to this version of connector? Will their jobs stop working until they do a full reset?

DoNotPanicUA replied:
There are three possible cases: 
1. Everything works as was before.
2. Sync skip part of the data
3. Sync fails

Case #1 is possible if you don't have any arrays. But usually, customers use BigQuery Denormalized destinations when they have complex types like arrays or records, and the usual destination can't normalize them. The result is entirely dependent on the user connector schema.

I haven't faced a connector update without backward compatibility. Shall we implement a feature that will force you to reset a destination if you upgraded the version without backward compatibility?

grishick replied:
The solution here is certainly not to force a reset of a destination after the upgrade - that's the last resort. We cannot release a new version of BQ Denormalized destination connector if the upgrade requires a reset, neither if the upgrade results in parts of data being skipped that were not skipped before. There needs to be better handling of the upgrade scenario here.

grishick replied:
before we can ship this, I need to understand the scope of disruption that this change will cause to existing customers and possible ways to avoid the disruption. As a starting point, please identify and describe the conditions that result in each of these behaviors post-upgrade and codify them as integration tests

1. Everything works as was before.
2. Sync skips part of the data
3. Sync fails

Once we have these conditions codified we'll have to look for solutions to minimize disruptions.

DoNotPanicUA replied:
#14440 investigation issue

edgao commented:
to clarify - this PR is to fix all 4 of those issues? Or will some of them still be open after this?

DoNotPanicUA commented:
@edgao 
I fix https://github.com/airbytehq/airbyte/issues/11109, https://github.com/airbytehq/airbyte/issues/14058, https://github.com/airbytehq/airbyte/issues/13634, and rollback https://github.com/airbytehq/airbyte/issues/5912

edgao commented on airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatter.java:
```diff
@@ -59,7 +59,7 @@ public DefaultBigQueryDenormalizedRecordFormatter(final JsonNode jsonSchema, fin
   @Override
   protected JsonNode formatJsonSchema(final JsonNode jsonSchema) {
-    var modifiedJsonSchema = formatAllOfAndAnyOfFields(namingResolver, jsonSchema);
+    var modifiedJsonSchema = jsonSchema.deepCopy(); // Issue #11166 is reopened formatAllOfAndAnyOfFields(namingResolver, jsonSchema);
```
I think this should reference https://github.com/airbytehq/airbyte/issues/5912, which is the issue that PR 11166 resolved.

DoNotPanicUA replied:
comment is updated

edgao commented on airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatter.java:
```diff
@@ -112,14 +113,17 @@ private void populateEmptyArrays(final JsonNode node) {
-          newNode.putArray(TYPE_FIELD).add("object");
-          newNode.putObject(PROPERTIES_FIELD).set(NESTED_ARRAY_FIELD, arrayNode);
+            originalNode.remove(ARRAY_ITEMS_FIELD);
+            final ObjectNode itemsNode = originalNode.putObject(ARRAY_ITEMS_FIELD);
```
checking my understanding:
* previously, a JsonNode representing an array schema would be replaced with a JsonNode representing an object schema, containing a `big_query_array` property whose value was the original schema
* the new logic is: the outer JsonNode is not wrapped inside an object, but it the items inside that array are, themselves, arrays, then they will still be wrapped in objects?

if that's accurate: wouldn't this have the same problem? Why do we need to do this wrapping at all?

https://github.com/airbytehq/airbyte/issues/11109 sounds like the user wants to recreate the source schema as closely as possible - this seems like we're basically pushing the problematic behavior down one level.

DoNotPanicUA replied:
BigQuery has limitations for arrays. You can't create an array of arrays.
![image](https://user-images.githubusercontent.com/30464745/179348042-10f13b74-9fda-4e52-9475-40d7eb5e8f37.png)
To avoid failure old implementation replaces all arrays by objects with the array field. 
As result, `Array of arrays` becomes `Object with array of objects with array`.
The new implementation does the schema transformation only when it's needed. We find cases when we have `array of arrays` and replace it with `array of objects with array`.

So, we change the incoming schema only when it's required.

edgao commented on airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatter.java:
```diff
@@ -171,7 +175,11 @@ protected JsonNode formatData(final FieldList fields, final JsonNode root) {
     if (!dateTimeFields.isEmpty() && !root.isNull()) {
-      BigQueryUtils.transformJsonDateTimeToBigDataFormat(dateTimeFields, (ObjectNode) root);
+      if (root.isArray()) {
+        root.forEach(jsonNode -> BigQueryUtils.transformJsonDateTimeToBigDataFormat(dateTimeFields, (ObjectNode) jsonNode));
```
does this work for double-nested arrays? E.g. if `root` looks like this:
```json
[
  [{"foo": "bar"}, {"foo": "baz"}]
]
```

DoNotPanicUA replied:
good point.
Fix is provided + few test cases:
![Screenshot from 2022-07-04 18-19-52](https://user-images.githubusercontent.com/30464745/177182961-fda70db6-f714-4f02-bedb-74cf2fb31ae4.png)

edgao reviewed (commented):
still reading through the rest of the PR, but have some questions the record formatter specifically

DoNotPanicUA commented on airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/formatter/DefaultBigQueryDenormalizedRecordFormatter.java:
```diff
@@ -59,7 +59,7 @@ public DefaultBigQueryDenormalizedRecordFormatter(final JsonNode jsonSchema, fin
   @Override
   protected JsonNode formatJsonSchema(final JsonNode jsonSchema) {
-    var modifiedJsonSchema = formatAllOfAndAnyOfFields(namingResolver, jsonSchema);
+    var modifiedJsonSchema = jsonSchema.deepCopy(); // Issue #11166 is reopened formatAllOfAndAnyOfFields(namingResolver, jsonSchema);
```
The feature is still needed and requires a fix. The fix is not so obvious and simple to be implemented as part of this PR.
I discussed with the @andriikorotkov and agreed to left the functionality as-is but disable it.

DoNotPanicUA commented on airbyte-integrations/connectors/destination-bigquery-denormalized/src/test-integration/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestinationTest.java:
```diff
@@ -250,7 +248,11 @@ void testWriteWithFormat() throws Exception {
 
     // Bigquery's datetime type accepts multiple input format but always outputs the same, so we can't
     // expect to receive the value we sent.
-    assertEquals(Set.of("2021-10-11T06:36:53"), extractJsonValues(resultJson, "updated_at"));
```
The hardcoded expected result as a string is the simplest solution, but it is also dangerous. 
Based on your local timezone your actual time will be different.
So, I replace it with a true comparison of the actual and expected values to make this test more stable.

DoNotPanicUA commented on airbyte-integrations/connectors/destination-bigquery-denormalized/src/test-integration/resources/testdata/dataArrays.json:
```diff
@@ -0,0 +1,9 @@
+  "simple_string": "simple string",
+  "array_1": [[1]],
+  "array_4": [[[4]]],
+  "array_5": [[[[5]]]]
```
added test case
![Screenshot from 2022-07-04 18-18-43](https://user-images.githubusercontent.com/30464745/177182781-4670bd12-fe00-410a-b203-d3a781486e0e.png)

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery-denormalized

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery-denormalized

grishick commented on airbyte-integrations/connectors/destination-bigquery-denormalized/src/test-integration/resources/testdata/expectedDataArrays.json:
```diff
@@ -0,0 +1,33 @@
+        {
+          "big_query_array": [
+            {
+              "big_query_array": [5]
```
then, we should have a test case with 15 levels and a test case with 16 levels to be sure that the connector behaves as expected for edge cases.

DoNotPanicUA replied:
I didn't cover these cases because it is not a limitation or specific to our implementation. Our implementation doesn't have any depth limit. So, the test of 3 nested arrays proves the algorithm's scalability is the same as for 15.
On the other hand, the test of 16 levels might be unstable because the BQ can change it at any moment, and our test will fail.

grishick replied:
> I didn't cover these cases because it is not a limitation or specific to our implementation. Our implementation doesn't have any depth limit. So, the test of 3 nested arrays proves the algorithm's scalability is the same as for 15. On the other hand, the test of 16 levels might be unstable because the BQ can change it at any moment, and our test will fail.

here's what I would like to confirm:
 - our connector can handle 15 levels of nesting
 - when source data has more levels of nesting than destination can support - we handle the error correctly and behavior is deterministic

DoNotPanicUA replied:
Follow-up issue #14668

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery-denormalized

zbrak commented:
Anything I may be able to contribute to help move this PR through? I have a fair amount waiting on our ability to cutover to the new `bigquery-denormalized` version.

Have been sitting on version `0.1.11` approaching a year due to the array surround problem.

DoNotPanicUA commented:
> Anything I may be able to contribute to help move this PR through? I have a fair amount waiting on our ability to cutover to the new `bigquery-denormalized` version.
> 
> Have been sitting on version `0.1.11` approaching a year due to the array surround problem.

Hi @zbrak 
Unfortunately, we can't release it as is so far. This change will affect existing cloud users. So, this PR probably will be on hold till we finalize our release strategy for such critical changes.
But I can release this branch version with postfix for your use. You could set this version on your instance only if you know the full version name, and it will not be visible to regular users.

@grishick
Don't you mind if I release this branch as `1.1.11b` version? We already practiced with such versions earlier.

zbrak commented:
> > Anything I may be able to contribute to help move this PR through? I have a fair amount waiting on our ability to cutover to the new `bigquery-denormalized` version.
> > Have been sitting on version `0.1.11` approaching a year due to the array surround problem.
> 
> Hi @zbrak Unfortunately, we can't release it as is so far. This change will affect existing cloud users. So, this PR probably will be on hold till we finalize our release strategy for such critical changes. But I can release this branch version with postfix for your use. You could set this version on your instance only if you know the full version name, and it will not be visible to regular users.
> 
> @grishick Don't you mind if I release this branch as `1.1.11b` version? We already practiced with such versions earlier.

Would this work if we're hosting our own instance @DoNotPanicUA ? We have a terraform deployment where we set the build version and helm chart version. But what you're suggesting sounds more isolated than that.

DoNotPanicUA commented:
> > > Anything I may be able to contribute to help move this PR through? I have a fair amount waiting on our ability to cutover to the new `bigquery-denormalized` version.
> > > Have been sitting on version `0.1.11` approaching a year due to the array surround problem.
> > 
> > 
> > Hi @zbrak Unfortunately, we can't release it as is so far. This change will affect existing cloud users. So, this PR probably will be on hold till we finalize our release strategy for such critical changes. But I can release this branch version with postfix for your use. You could set this version on your instance only if you know the full version name, and it will not be visible to regular users.
> > @grishick Don't you mind if I release this branch as `1.1.11b` version? We already practiced with such versions earlier.
> 
> Would this work if we're hosting our own instance @DoNotPanicUA ? We have a terraform deployment where we set the build version and helm chart version. But what you're suggesting sounds more isolated than that.

@zbrak
If you can go to the settings on your instance and pull the version for the BigQuery Denormalized (as you normally upgrade the version), I don't see a reason why it would not work for you.
Will be this version as latest for anyone who starts Airbyte? No. But anyone can get this version of the destination by pulling a specific version name.

grishick commented:
@DoNotPanicUA would it be possible to modify this PR in one of the following ways:

1. make the code check if the destination has "old" schema and fall back to the old behavior instead of failing the sync job
2. make the code check if the destination has "old" schema and modify the destination schema and data to match the new format

DoNotPanicUA commented:
> @DoNotPanicUA would it be possible to modify this PR in one of the following ways:
> 
> 1. make the code check if the destination has "old" schema and fall back to the old behavior instead of failing the sync job
> 2. make the code check if the destination has "old" schema and modify the destination schema and data to match the new format

@grishick 
Everything is possible :)
I've created [the issue](https://github.com/airbytehq/airbyte/issues/15363) for tracking.

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery-denormalized

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery-denormalized

DoNotPanicUA commented:
> @DoNotPanicUA would it be possible to modify this PR in one of the following ways:
> 
>     1. make the code check if the destination has "old" schema and fall back to the old behavior instead of failing the sync job
> 
>     2. make the code check if the destination has "old" schema and modify the destination schema and data to match the new format

@grishick 
I've implemented the workaround for the destination. While I'm working on tests for it, I would like to know your opinion about my solution.

1. I moved array transformation logic into separate classes. So, we now have `ModernArrayFormatter` and `ObsoleteArrayFormatter` with the new and old implementations.
2. When destination sync starts, we generate destination table schema using the `ModernArrayFormatter`. (As we usually do) 
3. Then, we get the real schema of the existing target table. If it is the same as we generated or the table doesn't exist(first run), we keep using the `ModernArrayFormatter` implementation.
But, if we have mismatch, we try to use `ObsoleteArrayFormatter`.

As a result, we let old connections proceed with their syncs without any resets, and we automatically start using the new processing for connections that are not affected, reset, or created after version upgrade.

grishick commented on airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java:
```diff
@@ -67,6 +78,30 @@ protected Function<String, String> getTargetTableNameTransformer(final BigQueryS
+    {
+      LOGGER.info("Target table already exists. Checking could we use the modern destination processing!");
+      com.google.cloud.bigquery.Schema existingTableSchema = existingTable.getDefinition().getSchema();
+      if (existingTableSchema != null && !existingTableSchema.equals(formatter.getBigQuerySchema())) {
```
What worries me about this comparison is whether there are conditions under which `existingTableSchema.equals(formatter.getBigQuerySchema())` evaluates to false even though the target table has been created by the new formatter. I looked at the implementation of `equals` method in `com.google.cloud.bigquery.Schema` and from what I can see, the way it evaluates the comparison is by comparing the protobuf representation of fields:

```
public boolean equals(Object obj) {
    return obj == this || obj instanceof Schema && Objects.equals(toPb(), ((Schema) obj).toPb());
  }


  TableSchema toPb() {
    TableSchema tableSchemaPb = new TableSchema();
    if (fields != null) {
      tableSchemaPb.setFields(fields.toPb());
    }
    return tableSchemaPb;
  }
```

`toPb` method of a field looks like following:
```
  TableFieldSchema toPb() {
    TableFieldSchema fieldSchemaPb = new TableFieldSchema();
    fieldSchemaPb.setName(name);
    fieldSchemaPb.setType(type.name());
    if (mode != null) {
      fieldSchemaPb.setMode(mode);
    }
    if (description != null) {
      fieldSchemaPb.setDescription(description);
    }
    if (policyTags != null) {
      fieldSchemaPb.setPolicyTags(policyTags.toPb());
    }
    if (getSubFields() != null) {
      List<TableFieldSchema> fieldsPb = Lists.transform(getSubFields(), TO_PB_FUNCTION);
      fieldSchemaPb.setFields(fieldsPb);
    }
    return fieldSchemaPb;
  }
```

based on this code, `existingTableSchema.equals(formatter.getBigQuerySchema())` will evaluate to `false` under any of the following conditions:

1. user added a field to the destination table
2. user added a description to a field in the destination table
3. user changed `NULLABLE`, `REQUIRED`  or `REPEATED` mode of a field
4. user set policy tags
5. user renamed field
6. user changed field type

of the 6 conditions listed above, conditions 1 - 4 should not cause us to fallback to `ObsoleteArrayFormatter` or to fail.

grishick commented on airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java:
```diff
@@ -67,6 +78,30 @@ protected Function<String, String> getTargetTableNameTransformer(final BigQueryS
+      com.google.cloud.bigquery.Schema existingTableSchema = existingTable.getDefinition().getSchema();
+      if (existingTableSchema != null && !existingTableSchema.equals(formatter.getBigQuerySchema())) {
+        ((DefaultBigQueryDenormalizedRecordFormatter) formatter).setArrayFormatter(new ObsoleteArrayFormatter());
+        LOGGER.warn("Existing target table has different structure with the new destination processing! Trying old implementation!!!");
```
nit: go easy on the exclamations

grishick commented on airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/formatter/arrayformater/ModernArrayFormatter.java:
```diff
@@ -0,0 +1,80 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ModernArrayFormatter implements ArrayFormatter {
```
nit: I don't think `Modern` adds value as a description here. Please rename to `DefaultArrayFormatter`

grishick commented on airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/formatter/arrayformater/ObsoleteArrayFormatter.java:
```diff
@@ -0,0 +1,61 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ObsoleteArrayFormatter extends ModernArrayFormatter {
```
nit: technically, this is not `obsolete`, because it is still being used, but `legacy`, so let's rename it to `LegacyArrayFormatter`

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery-denormalized

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery-denormalized

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery-denormalized

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery-denormalized

DoNotPanicUA commented:
### Manual testing results: 

| Case | Expected result | Actual result | Summary |
|---|---|---|---|
| [Incremental] New field to the end | Use default processing |||
| [Incremental] New field in the middle | Use default processing |||
| [Incremental] Field `not null` option change | Use default processing |||
| [Incremental] Add description to a field | Use default processing |||
| [Incremental] Add description to the table | Use default processing |||
| [Full overwrite] New field to the end | Use default processing | Use default processing. Table recreated by BQ | :heavy_check_mark: Use right implementation. :warning: BQ client erases all user modifications in this mode! |
| [Full overwrite] New field in the middle | Use default processing | Use default processing. Table recreated by BQ | :heavy_check_mark: Use right implementation. :warning: BQ client erases all user modifications in this mode! |
| [Full overwrite] Field `not null` option change | Use default processing | Use default processing. Table recreated by BQ | :heavy_check_mark: Use right implementation. :warning: BQ client erases all user modifications in this mode! |
| [Full overwrite] Add description to a field | Use default processing | Use default processing. Table recreated by BQ | :heavy_check_mark: Use right implementation. :warning: BQ client erases all user modifications in this mode! |
| [Full overwrite] Add description to the table | Use default processing | Use default processing. Table recreated by BQ | :heavy_check_mark: Use right implementation. :warning: BQ client erases all user modifications in this mode! |

grishick commented:
quick note: it would have been easier to discuss the specific test scenarios if they were numbered :)

Regarding second from the last scenario `[Full overwrite] Change field type` - if BQ client erases all user modifications then why would it fall back to legacy mode instead of using the new mode? I think that in full overwrite mode BQ client should always use the new implementation as long as it re-creates the destination table with correct field types.

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery-denormalized

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery

DoNotPanicUA commented:
/test connector=connectors/destination-bigquery-denormalized

grishick commented on airbyte-integrations/connectors/destination-bigquery-denormalized/src/main/java/io/airbyte/integrations/destination/bigquery/BigQueryDenormalizedDestination.java:
```diff
@@ -67,6 +82,119 @@ protected Function<String, String> getTargetTableNameTransformer(final BigQueryS
+  }
+
+  /**
+   * Compare field modes. Field can have on of three modes: NULLABLE, REQUIRED, REPEATED, null. Only
```
typo: "one of three modes"

DoNotPanicUA replied:
Fixed

DoNotPanicUA commented:
@grishick 
Followup ticket for the tests #16529
